kube-controllers/pkg/controllers/node/ipam.go

```diff
@@ -486,57 +546,195 @@ func (c *ipamController) onBlockDeleted(key model.BlockKey) {
 	delete(c.allBlocks, blockCIDR)
 	delete(c.nodesByBlock, blockCIDR)
 	delete(c.emptyBlocks, blockCIDR)
+
+	pool := c.poolsByBlock[blockCIDR]
+	delete(c.blocksByPool[pool], blockCIDR)
+	delete(c.poolsByBlock, blockCIDR)
+}
+
+func (c *ipamController) onPoolUpdated(kvp model.KVPair) {
+	pool := kvp.Value.(*apiv3.IPPool)
+	_, poolNet, err := cnet.ParseCIDR(pool.Spec.CIDR)
+	if err != nil {
+		log.WithError(err).Warnf("Unable to parse CIDR for IP Pool %s", pool.Name)
+		return
+	}
+
+	c.matchUnknownBlocksToPool(pool, poolNet)
```

Comment:
`matchUnknownBlocksToPool( ... )` not only matches, but also it updates.

If you change it from a method into a static func with simple in- and outputs, you can do the actual updates here in `onPoolUpdated( ... )`

Reply:
Thanks for the feedback, I expanded the logic into the pool update function, let me know if this addresses your point. The pool update and delete functions are now more symmetric as a nice side-effect.

I also addressed your other smaller comments, I marked them as resolved but please feel free to raise them again if I missed something.